arwen/host/asyncCall.go

```diff
@@ -0,0 +1,156 @@
+	if err != nil {
+		return err
+	}
+
+	_, err = host.ExecuteOnDestContext(callbackCallInput)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (host *vmHost) canExecuteSynchronously() bool {
+	// TODO replace with a blockchain hook that verifies if the caller and callee
+	// are in the same Shard.
+	runtime := host.Runtime()
+	blockchain := host.Blockchain()
+	asyncCallInfo := runtime.GetAsyncCallInfo()
+	dest := asyncCallInfo.Destination
+	calledSCCode, err := blockchain.GetCode(dest)
+
+	return len(calledSCCode) != 0 && err == nil
+}
+
+func (host *vmHost) createDestinationContractCallInput() (*vmcommon.ContractCallInput, error) {
+	runtime := host.Runtime()
+	sender := runtime.GetSCAddress()
+	asyncCallInfo := runtime.GetAsyncCallInfo()
+
+	argParser := runtime.ArgParser()
+	err := argParser.ParseData(string(asyncCallInfo.Data))
+	if err != nil {
+		return nil, err
+	}
+
+	function, err := argParser.GetFunction()
+	if err != nil {
+		return nil, err
+	}
+
+	arguments, err := argParser.GetArguments()
+	if err != nil {
+		return nil, err
+	}
+
+	gasLimit := asyncCallInfo.GasLimit
+	gasToUse := host.Metering().GasSchedule().ElrondAPICost.AsyncCallStep
+	if gasLimit <= gasToUse {
+		return nil, arwen.ErrNotEnoughGas
+	}
+	gasLimit -= gasToUse
+
+	contractCallInput := &vmcommon.ContractCallInput{
+		VMInput: vmcommon.VMInput{
+			CallerAddr:  sender,
+			Arguments:   arguments,
+			CallValue:   big.NewInt(0).SetBytes(asyncCallInfo.ValueBytes),
+			CallType:    vmcommon.AsynchronousCall,
+			GasPrice:    runtime.GetVMInput().GasPrice,
+			GasProvided: gasLimit,
+		},
+		RecipientAddr: asyncCallInfo.Destination,
+		Function:      function,
+	}
+
+	return contractCallInput, nil
+}
+
+func (host *vmHost) createCallbackContractCallInput(destinationVMOutput *vmcommon.VMOutput) (*vmcommon.ContractCallInput, error) {
+	metering := host.Metering()
+	runtime := host.Runtime()
+
+	arguments := destinationVMOutput.ReturnData
+	gasLimit := destinationVMOutput.GasRemaining
+	function := "callBack"
+
+	dataLength := host.computeDataLengthFromArguments(function, arguments)
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.AsyncCallStep
+	gasToUse += metering.GasSchedule().BaseOperationCost.DataCopyPerByte * uint64(dataLength)
+	if gasLimit <= gasToUse {
+		return nil, arwen.ErrNotEnoughGas
+	}
+	gasLimit -= gasToUse
+
+	sender := runtime.GetAsyncCallInfo().Destination
+	dest := runtime.GetSCAddress()
+
+	// Return to the sender SC, calling its callback() method.
+	contractCallInput := &vmcommon.ContractCallInput{
+		VMInput: vmcommon.VMInput{
+			CallerAddr:  sender,
+			Arguments:   arguments,
+			CallValue:   big.NewInt(0),
+			CallType:    vmcommon.AsynchronousCallBack,
+			GasPrice:    runtime.GetVMInput().GasPrice,
+			GasProvided: gasLimit,
+		},
+		RecipientAddr: dest,
+		Function:      function,
+	}
+
+	return contractCallInput, nil
+}
+
+func (host *vmHost) computeDataLengthFromArguments(function string, arguments [][]byte) int {
```

Comment:
Add unit test for this.

Reply:
Will be changed soon. A unit test will be written in the corresponding PR.